batch/batch/resource_usage.py

```diff
@@ -0,0 +1,108 @@
+import asyncio
+import pandas as pd
+import psutil
+import struct
+from typing import Optional
+
+from hailtop.utils import (
+    periodically_call,
+    retry_long_running,
+    time_msecs,
+    time_ns,
+)
+
+
+class ResourceUsageMonitor:
+    VERSION = 1
+
+    @staticmethod
+    def empty_df() -> pd.DataFrame:
+        return pd.DataFrame(
+            {
+                'time_msecs': pd.Series(dtype='int'),
+                'cpu_usage': pd.Series(dtype='float'),
+                'memory_in_bytes': pd.Series(dtype='int'),
+            }
+        )
+
+    @staticmethod
+    def decode_to_df(data: bytes) -> pd.DataFrame:
+        offset = 0
+        decoded_data = []
+
+        (version,) = struct.unpack_from('>i', data, offset)
+        offset += 4
+        assert version == ResourceUsageMonitor.VERSION, version
+
+        while offset < len(data):
+            timestamp_msecs, cpu_usage, memory_in_bytes = struct.unpack_from('>qdq', data, offset)
+            decoded_data.append(
+                {'time_msecs': timestamp_msecs, 'cpu_usage': cpu_usage, 'memory_in_bytes': memory_in_bytes}
+            )
+            offset += 24
```

Comment:
I think the main issue is that we're spending too much time in this Python while loop and not enough time in optimized C code.

Use the built-in repetition system (described under [Format Characters](https://docs.python.org/3/library/struct.html#format-characters)):
> A format character may be preceded by an integral repeat count. For example, the format string '4h' means exactly the same as 'hhhh'.

Super tiny minor thing that probably doesn't matter, but let's use a long for the version. That ensures everything is aligned on 8-byte boundaries.

Reply:
You should be able to reshape efficiently with numpy:
```python3
n_records = (len(data) - HEADER_LENGTH) // 3
np.array(data).reshape((n_records, 3))
```

Reply:
Using numpy helped a lot!